terraform/gen/referencedocs.go.tpl

```diff
@@ -67,6 +67,22 @@ provider "teleport" {
   root_ca_path = "tf.ca"
 }
 ```
+
+## Provider resource versioning
+
+Teleport resource APIs are versioned. Based on the provided version, Teleport will
+set different defaults. The Terraform provider might not see the changes caused
+by a resource version increase as version is used to set defaults on creation.
+However those changes will take effect as soon as Terraform will need to update
+the resource for any other reason.
+
+Since Teleport 15, you MUST set the version on each resource. This is not
+enforced on previous Teleport provider versions but we highly recommend doing
+so. If it's not specified, Terraform will pick the latest one by default. When
+this happens, you will face unexpected changes when doing major Terraform
+provider upgrades, potentially introducing new resource versions
+and changing the behaviour of your existing TF code.
```

Comment:
Worth mentioning that when dealing with roles, it can impact access rules when upgrading the provider across major versions

Reply:
I don't want to sound like a broken record but IMO we do also need some central reference for what role versions can be used, what is current, what the difference is between them, etc. We'll probably get a lot of questions about "what role version should I use" otherwise (more than we usually get)

Reply:
I wrote the tests, and I have bad news: we built a system with hysteresis 😭 .

A role upgraded to v6 from v5, and a role created from v6 with the same spec won't have the same rules.

Resources have their defaults set on initial creation, then never again. If you create a resource v5, you will have the v5 defaults, even after you upgrade to v6.

There's an interaction between Teleport resources without clear field ownership and how Terraform measures the existing state to do drift detection. Teleport edits the spec server-side, losing the original user intention. Then TF imports this amended resource back in its state, importing all the defaults and keeping them ad vitam.

The conclusion is that upgrading a resource between versions is not safe. As re-creating it from scratch will not lead to the same result. This also explains why so few people complained about their role being broken after upgrading to Teleport 13.